torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -856,9 +1016,62 @@ void SegmentCandidateFinder::findSegments() {
     mergeNodes();
   }
 
+  finalMerge();
+
   finalize();
 }
 
+void SegmentCandidateFinder::finalMerge() {
+  BackwardReachability reachability(segmented_fusion_.get());
+
+  bool merged_nodes = true;
+  while (merged_nodes) {
+    // One merge per iteration
+    for (auto g : groups()) {
+      // populate consumers
+      std::unordered_map<SegmentedGroup*, SegmentedEdge*> edge_map;
+      for (auto c : g->consumer_edges) {
+        edge_map.insert({c->to, c});
+      }
+      std::vector<SegmentedGroup*> all_consumers;
+      std::transform(
+          edge_map.begin(),
+          edge_map.end(),
+          std::back_inserter(all_consumers),
+          [](auto& it) { return it.first; });
+
+      for (auto c : all_consumers) {
+        if (!reachability.reachesAny(c, all_consumers) &&
```

Comment:
reachability seems ambiguous to me. This is checking if in c there are chains through producer that reaches any group in all_consumers.
Since all consumers are consumers of c, shouldn't this actually be `!reachability.reachesAny(g, all_consumers)`? If this is already a DAG how could a consumer of c be in its producer path?

Reply:
If I want to see if a `producer` can be merged with a `consumer` I want to make sure that no consumer of producer reaches producer, or no producer of consumer reaches producer. This seems to be checking that no producer of c reaches consumers of c, or I read this wrong which could also be the case.

Reply:
Renamed `reachability` to `producer_check` to avoid confusion.

Reply:
Updated comment in fusion_segmenter.h to avoid confusion. 

There could be a few ways to check the merging condition. This implementation is based on Theorem 4.1 of the paper we used for segmentation (Herrmann et al.). In the case of checking if a producer and a consumer can be merged into one node. I believe based on Theorem 4.1, it is enough to check if any other consumer of the same producer also produces the consumer. e.g.

say **a** produces **b**, **c**, **d**, want to check if **a** and **b** can merge. This implementation checks 

!(**c** produces **b**) and !(**d** produces **b**) or as written in the code:
`!isConsumerOfAny(b,{c,d})`

Reply:
Thanks for the feedback and please let me know if there are more questions or suggestions.